cpu/esp8266/syscalls.c

```diff
@@ -18,342 +18,573 @@
+
+    free((void*)*lock);
+    *lock = 0;
 }
 
-void IRAM syscalls_init (void)
+void IRAM _lock_acquire(_lock_t *lock)
 {
-}
+    assert(lock != NULL);
 
-#else   /* MODULE_ESP_SDK */
-/*
- * To use the same heap as SDK internally does, SDK memory management
- * functions have to be replaced by newlib memory functions. In that case,
- * the _malloc_lock/_unlock functions have to be defined. Memory management
- * functions of ETS are not used and have not to be considered here.
- */
+    /* if the lock data structure is still not allocated, initialize it first */
+    if (*lock == 0) {
+        _lock_init(lock);
+    }
 
-void* IRAM pvPortMalloc (size_t size, const char *file, unsigned line)
-{
-    (void)file;
-    (void)line;
+    /* if scheduler is not running, we have not to lock the mutex */
+    if (sched_active_thread == NULL) {
+        return;
+    }
 
-    return malloc (size);
+    assert(!irq_is_in());
+    mutex_lock((mutex_t*)*lock);
 }
 
-void IRAM vPortFree (void *ptr, const char *file, unsigned line)
+void IRAM _lock_acquire_recursive(_lock_t *lock)
 {
-    (void)file;
-    (void)line;
+    assert(lock != NULL);
+
+    /* if the lock data structure is still not allocated, initialize it first */
+    if (*lock == 0) {
+        _lock_init_recursive(lock);
+    }
+
+    /* if scheduler is not running, we have not to lock the rmutex */
+    if (sched_active_thread == NULL) {
+        return;
+    }
 
-    free (ptr);
+    assert(!irq_is_in());
+    rmutex_lock((rmutex_t*)*lock);
 }
 
-void* IRAM pvPortCalloc (size_t nmemb, size_t size, const char *file, unsigned line)
+int IRAM _lock_try_acquire(_lock_t *lock)
 {
-    (void)file;
-    (void)line;
+    assert(lock != NULL);
 
-    void *ptr = malloc (nmemb*size);
-    if (ptr) {
-        memset (ptr, 0x0, nmemb*size);
+    /* if the lock data structure is still not allocated, initialize it first */
+    if (*lock == 0) {
+        _lock_init(lock);
     }
-    return ptr;
-}
 
-void* IRAM pvPortZalloc (size_t size, const char *file, unsigned line)
-{
-    (void)file;
-    (void)line;
+    /* if scheduler is not running, we have not to lock the mutex */
+    if (sched_active_thread == NULL) {
+        return 0;
+    }
 
-    void *ptr = malloc (size);
-    if (ptr) {
-        memset (ptr, 0x0, size);
+    if (irq_is_in()) {
+        return 0;
     }
-    return ptr;
+
+    return rmutex_trylock((rmutex_t*)*lock);
 }
 
-void* IRAM pvPortRealloc (void *ptr, size_t size, const char *file, unsigned line)
+int IRAM _lock_try_acquire_recursive(_lock_t *lock)
 {
-    (void)file;
-    (void)line;
+    assert(lock != NULL);
 
-    return realloc (ptr, size);
+    /* if the lock data structure is still not allocated, initialize it first */
+    if (*lock == 0) {
+        _lock_init_recursive(lock);
+    }
+
+    /* if scheduler is not running, we have not to lock the rmutex */
+    if (sched_active_thread == NULL) {
+        return 0;
+    }
+
+    return rmutex_trylock((rmutex_t*)*lock);
```

Comment:
@kaspar030 As the responsible maintainer for the core, I would like to clarify with you the following question before the PR is merged.

Even though it seems from comprehensive tests that this `newlib` lock function isn't used in any ISR, I would like to be sure that it  wouldn't be a problem if it would be used from in an ISR for any reason.

From my point of view, if an ISR calls `rmutex_trylock` here and the mutex is already locked, then it is locked by the interrupted thread. That is, for the `rmutex` it seems like the same thread is locking the mutex again (case 2: mutex is held by me (relock)). In other words, case 1 (mutex is not held by me) can't happen in an ISR. Right? If so, there no additional ISR guarding mechanism necessary.

Reply:
first of all, IMO, the two checks whether the mutex is initialized an whether the scheduler is running, can they be handled earlier in the boot process? The former hits exactly one time after boot, but has impact on all later calls. Similar with the check whether the scheduler is running.





> From my point of view, if an ISR calls `rmutex_trylock` here and the mutex is already locked, then it is locked by the interrupted thread.

Nope. If thread A holds the rmutex, thread B gets scheduled for any reason (something unrelated), then an ISR hits, sched_active_thread will be thread B, but the rmutex is held by A. Thus if the isr calls rmutex_lock(), case 1 hits.

Reply:
> Nope. If thread A holds the rmutex, thread B gets scheduled for any reason (something unrelated), then an ISR hits, sched_active_thread will be thread B, but the rmutex is held by A. Thus if the isr calls rmutex_lock(), case 1 hits.

Oh yeah, of course, the running thread being interrupted has not to be the holder of this mutex. Could be another thread that is not running because it is blocked, e.g., by another mutex. I was too fixated on the fact that only the thread could run, which also holds the Mutex.

Reply:
> first of all, IMO, the two checks whether the mutex is initialized an whether the scheduler is running, can they be handled earlier in the boot process?

Unfortunatly not really since the `newlib` calls `_lock_try_acquire` without calling `_lock_init` and I don't know what mutexes these are. And, first `_lock_try_acquire` calls appear before the scheduler is running, but not all of them.

Reply:
> Unfortunatly not really since the `newlib` calls `_lock_try_acquire` without calling `_lock_init` before and I don't know what mutexes these are. And, first `_lock_try_acquire` calls appear before the scheduler is running, but not all of them.

ok (for now. :wink).